pgjdbc/src/main/java/org/postgresql/core/QueryExecutorCloseAction.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The action performs connection cleanup, so it is properly terminated from the backend
+ * point of view.
+ * Implementation note: it should keep only the minimum number of object references
+ * to reduce heap usage in case the user abandons connection without closing it first.
+ */
+public class QueryExecutorCloseAction implements Closeable {
+  private static final Logger LOGGER = Logger.getLogger(QueryExecutorBase.class.getName());
+  private @Nullable PGStream pgStream;
+
+  public QueryExecutorCloseAction(PGStream pgStream) {
+    this.pgStream = pgStream;
+  }
+
+  public boolean isClosed() {
+    return pgStream == null;
+  }
+
+  public void abort() {
+    PGStream pgStream = this.pgStream;
+    if (pgStream == null) {
+      // The connection has already been closed
+      return;
+    }
+    try {
+      LOGGER.log(Level.FINEST, " FE=> close socket");
+      pgStream.getSocket().close();
+    } catch (IOException e) {
+      // ignore
+    }
+    this.pgStream = null;
+  }
+
+  @Override
+  public void close() throws IOException {
+    LOGGER.log(Level.FINEST, " FE=> Terminate");
+    PGStream pgStream = this.pgStream;
+    if (pgStream == null) {
+      // The connection has already been closed
+      return;
+    }
+    this.pgStream = null;
```

Comment:
is there risk of the close method being called concurrently?
if so, it would seem assigning pgStream to null should be CAS to avoid 2 threads trying to do the work below.